staging/src/k8s.io/kubectl/pkg/cmd/auth/whoami.go

```diff
@@ -0,0 +1,188 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/spf13/cobra"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	authenticationv1alpha1 "k8s.io/api/authentication/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/client-go/kubernetes"
+	authenticationv1alpha1client "k8s.io/client-go/kubernetes/typed/authentication/v1alpha1"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/scheme"
+	"k8s.io/kubectl/pkg/util/templates"
+)
+
+// WhoAmIFlags directly reflect the information that CLI is gathering via flags.  They will be converted to Options, which
+// reflect the runtime requirements for the command.  This structure reduces the transformation to wiring and makes
+// the logic itself easy to unit test.
+type WhoAmIFlags struct {
+	RESTClientGetter genericclioptions.RESTClientGetter
+	PrintFlags       *genericclioptions.PrintFlags
+
+	genericclioptions.IOStreams
+}
+
+// NewWhoAmIFlags returns a default WhoAmIFlags.
+func NewWhoAmIFlags(restClientGetter genericclioptions.RESTClientGetter, streams genericclioptions.IOStreams) *WhoAmIFlags {
+	return &WhoAmIFlags{
+		RESTClientGetter: restClientGetter,
+		PrintFlags:       genericclioptions.NewPrintFlags("").WithTypeSetter(scheme.Scheme),
+		IOStreams:        streams,
+	}
+}
+
+// AddFlags registers flags for a cli.
+func (flags *WhoAmIFlags) AddFlags(cmd *cobra.Command) {
+	flags.PrintFlags.AddFlags(cmd)
+}
+
+// ToOptions converts from CLI inputs to runtime inputs.
+func (flags *WhoAmIFlags) ToOptions(ctx context.Context, args []string) (*WhoAmIOptions, error) {
+	w := &WhoAmIOptions{
+		ctx:       ctx,
+		IOStreams: flags.IOStreams,
+	}
+
+	clientConfig, err := flags.RESTClientGetter.ToRESTConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	clientSet, err := kubernetes.NewForConfig(clientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	w.authClient = clientSet.AuthenticationV1alpha1()
+
+	if !flags.PrintFlags.OutputFlagSpecified() {
+		w.printTable = true
+		w.printer = printers.NewTablePrinter(printers.PrintOptions{})
+	} else {
+		w.printer, err = flags.PrintFlags.ToPrinter()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return w, nil
+}
+
+// WhoAmIOptions is the start of the data required to perform the operation. As new fields are added,
+// add them here instead of referencing the cmd.Flags()
+type WhoAmIOptions struct {
+	authClient authenticationv1alpha1client.AuthenticationV1alpha1Interface
+	ctx        context.Context
+
+	printer    printers.ResourcePrinter
```

Comment:
`PrintObj printers.ResourcePrinterFunc` see my previous comment.

Reply:
I migrate to using `ResourcePrinterFunc` here.